WIP:  Fix Calico install for BGP

catblade opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Add in actual nodes instead of all inventory for bgp routes when using BGP per-node scope.  Otherwise, half the links get dropped.
Add in Calico ability to advertise kube_service_addresses in the event that we are doing BGP at the config level.  Otherwise, Calico never takes over the config and app advertising will not work.

Tested on our little cluster with a switch.  Do not see infrastructure in kubepsray proper to test this type of config (needs switch configuration to test).  We have an iperf test that works with these changes.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #
https://github.com/kubernetes-sigs/kubespray/issues/6845

**Special notes for your reviewer**:
You may wish to talk to me directly in how to set a switch for testing.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE.
```

floryut commented:
/ok-to-test
/approve

floryut commented:
@catblade btw, Thank you for the PR ! Could you please sign [CLA](https://git.k8s.io/community/CLA.md#the-contributor-license-agreement) ?

catblade commented:
@floryut  I have attempted to sign the CLA, but their form never reached my inbox.  I tried to send a support ticket to them, and that email also failed... I'll see if it resolves in the next day.

marlow-h commented:
@floryut I have signed.

catblade commented:
@floryut Please do not merge yet.  There is still a bug left.  Will move to WIP until done.

mirwan commented on roles/network_plugin/calico/tasks/install.yml:
```diff
@@ -331,11 +332,11 @@
       "spec": {
         "asNumber": "{{ item.as }}",
-        "node": "{{ inventory_hostname }}",
+        "node": "{{ item.node }}",
```
`spec.node` is supposed to match the node's the value of `metadata.name` as defined in the previous task (which takes the value of `inventory_hostname`).
There is no need to define a "node" key in `peers` items.
`peers` should be defined in hostvars as it is a "per node peering"

catblade replied:
@mirwan  The above lack of nodes being added had all the peers for calico being added, so half the routes were failing.  Should only be adding the per-node.  I did fix for the generic case.

mirwan replied:
I think this is because you defined `peers` in groupvars whereas it should be defined in hostvars (where each `peers` list is different from one to another) am I right ?

catblade replied:
The peers in the examples are defined in the k8s-net-calico.yml in the group_vars/k8s-cluster/ directory.  I followed this example.  https://gitlab.staging.cncf.ci/cncf/kubespray/tree/master/docs/calico_peer_example
but added nodes.

Also not handled are nodeSelectors.

catblade commented:
@mirwan  I've added example files (for clarity) and added in the ability to do nodeSelector instead of node if desired.  Both have been tested with my config and switch.  I do not know what policy is to add in errors in the event someone tries to use both, so have it not entering in node reflectors if one of the three combinations is not followed (IE with nodeSelector ONLY, node ONLY, or without either).

catblade commented on docs/calico_peer_example/peer_by_node.yml:
```diff
@@ -0,0 +1,8 @@
+# peers:
+#   - router_id: "10.99.0.34"
+#     as: "65xxx"
+#     node: "nodename1"
```
I don't see anywhere in kubespray that properly sets up the peers for bgp configuration-the bgp configuration is done through the groupvars/k8s-cluster/k8s-net-calico.yml configuration.  This is a calico-specific configuration.

I am following the documentation given here:
https://docs.projectcalico.org/networking/bgp

and mimicking the current structure in kubespray.

mirwan replied:
groupvars/k8s-cluster/k8s-net-calico.yml holds global calico parameters among which `peer_with_router` defines whether or not external peering is enabled whatever the actual peers are (as they are specific to the ecosystem kubespray is run onto).
How to define peers in calico documentation is clear so it is more of how to leverage ansible to have the variables correctly defined (through group_vars and/or hostvars) to manage this.
I think that the "Optional : Defining BGP peers" section in docs/calico.md explains how to define those peers but maybe clarifications are required.

mirwan commented on roles/network_plugin/calico/tasks/install.yml:
```diff
@@ -347,3 +348,30 @@
     - inventory_hostname in groups['k8s-cluster']
+    - item.nodeSelector is not defined
+
+- name: Calico | Configure peering with router(s) via nodeSelectors
```
@catblade I think you might take a look at how the global peers are defined, adding the same condition so that the task is run only on the first master would do the trick.
And maybe add a `selectattr('nodeSelector','defined')` filter in the loop

mirwan commented:
@catblade As discussed on slack, could you make this PR only fix the clusterIP advertizement issue (btw removing CALICO_ADVERTISE_CLUSTER_IPS envvar from daemonset would be nice) and open another one to improve the node peering ?

catblade commented:
I'm going to close this for now and come back with it broken up.  Thanks @mirwan  for talking me through some of this.

catblade commented:
/rebase

catblade commented:
Closing and re-opening.  This fork became a mess and I only want to modify two files.
